Drag drop console cells into notebook

Madhu94 opened the pull request:
Fixes #4847

ian-r-rose commented:
Thanks for tackling this @Madhu94! Is it ready for review yet? I am trying it out and having a hard time getting the console cell drag to work.

Madhu94 commented:
Sorry, I didn't notice the above update. @ian-r-rose Could you try it out now? I wouldn't say it is ready for review - some  tests were failing and I couldn't figure out if I broke it. I will check those now

ian-r-rose commented:
I'd say don't worry about the tests too much unless they look specifically related to your functionality -- we are still working out some problems with the test suite. Better to get the functionality where you want it, and we can address the tests later.

I was just playing with this, and I am still getting some errors when I try to drag the cells:
```javascript
ReferenceError: event is not defined[Learn More] cellDragUtils.js:76
detectTargetArea
cellDragUtils.js:76
_evtMouseDown
widget.js:1160
handleEvent
widget.js:849
```

Madhu94 commented:
This was the error I got too, when you said earlier that it didn't work. I had fixed it later. I had squashed the fix commit into the original one and force pushed. Can you check if you pulled the latest changes from this branch? (you probably have, just confirming).

Madhu94 commented:
The test cases do not seem related, seems like a path resolution issue ?

ian-r-rose commented:
Ah, you are absolutely right @Madhu94, my apologies! I messed up the fetch-after-force-push.

ian-r-rose commented on packages/cells/src/cellDragUtils.ts:
```diff
@@ -0,0 +1,209 @@
+  return cellIndex;
+}
+
+export type ICellTargetArea = 'input' | 'prompt' | 'cell' | 'notebook';
```
This smells off to me. I think that, in general, the `@jupyterlab/cells` package shouldn't have to know about the context in which a cell is embedded. In this case, we are leaking the idea of a `'notebook'` into this package.

I don't think that the `detectTargetArea` code is so long that we can't have specific versions for the notebook and the console, especially since they have different behaviors. Then we wouldn't need to export a pretty specialized public interface.

Alternatively, we could remove the `'notebook'` option and add an `'unknown'` option, and the notebook drag handler could just assume it's in the notebook if it is `'unknown'`. What do you think @Madhu94?

ian-r-rose commented on packages/cells/src/cellDragUtils.ts:
```diff
@@ -0,0 +1,209 @@
+function findCell(
+  node: HTMLElement,
+  cells: IterableOrArrayLike<Cell>,
+  isCellNode: (node: HTMLElement) => boolean
```
This is a nice function reuse. I wonder if we should provide a default `isCellNode` that just checks for the `jp-Cell` class so that this function works well in the absence of a more specific filtering function. Not a huge deal either way, though.

Madhu94 commented on packages/cells/src/index.ts:
```diff
@@ -5,6 +5,7 @@
 
 import '../style/index.css';
 
+export {default as CellDragUtils, ICellTargetArea} from './cellDragUtils';
```
Do we intend to continue usage of namespaces? I've heard it isn't recommended, correct me if I'm wrong. E.g http://artsy.github.io/blog/2017/11/27/Babel-7-and-TypeScript/ (see section number 5)

ian-r-rose replied:
We still use namespaces pretty prolifically throughout the JupyterLab codebase. As far as I know, we haven't had any conversations about discouraging their usage.

jasongrout replied:
I took that online reference to mean that the Babel ts plugin has a limitation in that it can't handle namespaces, not that they are discouraged in general ts code (we aren't using Babel).

Madhu94 replied:
I see. One more question - do  they provide any advantages over plain ES6  modules ?

Madhu94 commented:
@ian-r-rose I've taken care of the review comments.  Let me know if there is something else that can be done.

Madhu94 commented:
Hi, any updates here?

ian-r-rose commented on packages/console/src/widget.ts:
```diff
@@ -373,10 +389,118 @@ export class CodeConsole extends Widget {
+   * Start a drag event
+   */
+  private _startDrag(index: number, clientX: number, clientY: number) {
+    const cellModel = this._focussedCell.model as ICodeCellModel;
```
Are we certain that `this._focussedCell` can never be `null` here? Perhaps we should double check.

jasongrout replied:
Just noticed that "focused" should have one s. Thanks!

ian-r-rose replied:
Hah, I almost wrote the same thing, but then I looked it up, and apparently two esses is a valid alternate.

jasongrout replied:
Oh, wow, TIL about this! According to https://en.wiktionary.org/wiki/Talk:focus (which cites OED) and https://english.stackexchange.com/questions/4791/focussed-or-focused-rules-for-doubling-the-last-consonant-when-adding-ed, "focused" is preferred these days, at least in US and British English.

Madhu94 replied:
I think it wouldn't be null if I handle the case you noticed above - that cellIndex won't be -1. What  do you think?

ian-r-rose replied:
I think that's right as long as `_startDrag` is only called from that code path (which,to be sure, is the case right now). I'm just trying to think defensively, but it's not crucial here.

ian-r-rose commented:
Thanks @Madhu94 (and thanks for being persistent)!

Madhu94 commented:
Thanks for the help @ian-r-rose ! I'll be more prompt on the review comments next time, keeping a PR alive for weeks is hard.

ian-r-rose commented:
No, the slow review cycle is entirely my fault, thanks again!
